### PR TITLE
Feat: podman - use systemd service to manage registry

### PR DESCRIPTION
**Pull request by ginomcevoy**
This PR is for a feature request to use systemd to manage the podman registry. The goal is to support an active/passive configuration for the registry in a HA cluster environment.

The original implementation works and would be usable in an active/active scenario as the README proposes. However there is a detail: when using this approach, each container will get its own volume mount (see https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/high_availability_add-on_reference/s1-containers-haar#s3-bundlestorage-properties-HAAR , I could not find this documentation for RHEL 8), meaning that the volume data is not shared by the instances. This may be fine for some services, but for registry I think it is best to push images once and have them available on the whole HA cluster.

With this PR, the registry could be managed by the high_availability role like this:

```
    - group: registry
      resources:
        - id: fs-registry
          type: Filesystem
          arguments: "device='/dev/registry' directory='{{ podman_local_registry_dir }}' fstype='ext4'"
        - id: vip-registry
          type: IPaddr2
          arguments: "ip={{ podman_local_registry_host }} cidr_netmask=255.255.0.0"
        - id: service-registry
          type: systemd:registry
          
```

(I did not consider adding this last bit to the podman README when I opened the PR, what do you think?)

Note: I have already tested the PR with a very similar HA configuration in a virtual cluster, and it works (registry can be relocated to a different node). The difference in the test is that I declared podman_local_registry_host with a variable from network.yml, and the pcs resource used the same variable definition. Also there was no Filesystem resource, but this example  follows the high_availability README.

### CODE REVIEW

**Comment by ginomcevoy** [comment]
Pinging @strus38 for review.

**Review by strus38 — APPROVED** [review]

**Comment by strus38** [comment]
Approved. Thanks!

**Comment by johnnykeats** [comment]
@strus38 You are podman role maintainer. Are you able to merge the PR to master?

**Comment by johnnykeats** [comment]
Please add a note here if you wish a new release after merge of podman and grafana PRs.

**Comment by strus38** [comment]
@johnnykeats - I don't see any button for me to merge!

**Comment by ginomcevoy** [comment]
Given that this PR has not been merged yet, can I add a commit to it?
Today I found a bug during a review of this role and I think I know how to improve the systemd service.

**Comment by oxedions** [comment]
@strus38 ok, now you should see the button (I hope).
@strus38 I let you check with @ginomcevoy

**Comment by strus38** [comment]
YEs please fix the bug before i merge

**Comment by ginomcevoy** [comment]
Brief explanation of the fixes:

- Sometimes the container would not be removed after stopping the service, so the service would not start again. Using the "podman generate systemd" template as a basis for the service fixes the issue.
- After migrating the service to another node in a HA scenario, for some reason the registry would stop working with "Connection refused". This got fixed after I enabled the port in the firewall. Not sure how it was working in the first place...

The service is performing better now and I think the PR is ready to merge. I will still verify on a new installation from scratch tomorrow.

**Comment by ginomcevoy** [comment]
OK so there is definitely an issue with podman and firewalld:
https://github.com/containers/podman/issues/5431

This affects all containers, not just registry. Here is an example of the problem:

1. Start the registry
```
# systemctl start registry
# curl --noproxy '*' -k -X GET http://10.31.0.57:5000/v2/_catalog
{"repositories":[]}
```

2. Reload the firewall. The containers will lose connectivity:

```
# firewall-cmd --reload
# curl --noproxy '*' -k -X GET http://10.31.0.57:5000/v2/_catalog
(hangs or gets "Connection refused")
```

First I was getting "Connection refused" after moving the registry from one management node and back using pacemaker, and this was without reloading firewall explicitly. Maybe that's a different issue. Now I only get a hanging service, and consistently.

The explanation is that podman is adding a rule to the firewall:
```
firewall-cmd --info-zone trusted
trusted (active)
  target: ACCEPT
  icmp-block-inversion: no
  interfaces:
  sources: 10.88.0.10/32
  services:
  ports:
  protocols:
  forward: no
  masquerade: no
  forward-ports:
  source-ports:
  icmp-blocks:
  rich rules:
```

When the firewall reloads, this change ("sources" field) is gone and containers lose connectivity. 

These are known fixes:
- re-create the container (of course we don't want this)
- run the command "podman network reload --all" after the firewall is reloaded

These may or may not work, I have not tested them yet:
- This script that runs the command above: https://github.com/containers/podman/issues/5431#issuecomment-899538120
- Using a rootless network and slirp4netns (suggested in the podman issue)

These do not work:
- Add the network 10.88.0.0/16 to the trusted zone, I guess that podman makes additional changes to the firewall.
- Add the cni-podman0 interface to the firewall, same as the previous one.

What do you think @strus38 ?

**Comment by strus38** [comment]
https://blog.christophersmart.com/2021/01/31/podman-volumes-and-selinux/

Le ven. 10 déc. 2021 à 18:12, Giacomo Mc Evoy ***@***.***> a
écrit :

**Comment by ginomcevoy** [comment]
Hello @strus38 I think SELinux is a matter that we will have to consider in the near future. Some containers will simply not work with SELinux enabled without the appropriate permissions, and the blog post can show us how we could secure access to the registry files so that other containers cannot modify them.

But this particular issue about the firewall is independent of SELinux, it shows up even in permissive mode. I think for now I will add a comment to the podman README with the "podman network reload --all"  workaround.

**Pull request merged**
